### ISSUE TITLE
sudo password fail

### ISSUE THREAD

**Issue by Bystroushaak** [Fizzadar/pyinfra#398]
**Describe the bug**
Pyinfra fails with traceback when trying to use sudo password for apt update.

**To Reproduce**

```python
from pyinfra.operations import apt


SUDO=True
USE_SUDO_PASSWORD=True


apt.update()
```

Run with

```
pyinfra  kitakitsune_inventory.py update_vps.py
```

**Meta**
+ Include output of `pyinfra --support`.

```
--> Support information:

    If you are having issues with pyinfra or wish to make feature requests, please
    check out the GitHub issues at https://github.com/Fizzadar/pyinfra/issues .
    When adding an issue, be sure to include the following:

    System: Linux
      Platform: Linux-4.15.0-20-generic-x86_64-with-LinuxMint-19.1-tessa
      Release: 4.15.0-20-generic
      Machine: x86_64
    pyinfra: v1.0.4
    Executable: /home/bystrousak/.local/bin/pyinfra
    Python: 2.7.17 (CPython, GCC 7.5.0)

```

Tested against newly created VPS with ubuntu 20.4.

+ How was pyinfra installed (source/pip)?
+ Include pyinfra-debug.log (if one was created)

```
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra_cli/main.py", line 220, in main
    _main(*args, **kwargs)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra_cli/main.py", line 576, in _main
    run_ops(state, serial=serial, no_wait=no_wait)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/operations.py", line 351, in run_ops
    _run_single_op(state, op_hash)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/operations.py", line 317, in _run_single_op
    if not greenlet.get():
  File "src/gevent/greenlet.py", line 769, in gevent._gevent_cgreenlet.Greenlet.get
  File "src/gevent/greenlet.py", line 364, in gevent._gevent_cgreenlet.Greenlet._raise_exception
  File "src/gevent/greenlet.py", line 854, in gevent._gevent_cgreenlet.Greenlet.run
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/operations.py", line 129, in _run_server_op
    **executor_kwargs
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/host.py", line 126, in run_shell_command
    return self.executor.run_shell_command(self.state, self, *args, **kwargs)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 258, in run_shell_command
    put_file=put_file,
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/util.py", line 142, in get_sudo_password
    put_file(state, host, SUDO_ASKPASS_EXE, filename)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 427, in put_file
    _put_file(host, filename_or_io, remote_filename)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 381, in _put_file
    sftp.putfo(file_io, remote_location)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 714, in putfo
    with self.file(remotepath, "wb") as fr:
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 372, in open
    t, msg = self._request(CMD_OPEN, filename, imode, attrblock)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 813, in _request
    return self._read_response(num)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 865, in _read_response
    self._convert_status(msg)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 896, in _convert_status
    raise IOError(errno.EACCES, text)
IOError: [Errno 13] Permission denied
```

+ Consider including output with `-vv` and `--debug`.

```
bystrousak:~/Plocha/Syncthing/c0d3z/devops/vps,1$ pyinfra --debug kitakitsune_inventory.py update_vps.py 
    [pyinfra_cli.main] Checking potential directory: 
    [pyinfra_cli.main] Checking potential directory: 
    [pyinfra_cli.main] Checking potential directory: 
    [pyinfra_cli.main] Checking potential directory: 
--> Loading config...
--> Loading inventory...
    [pyinfra_cli.inventory] Creating fake inventory...

--> Connecting to hosts...
    [pyinfra.api.connectors.ssh] Connecting to: 81.2.255.191 ({u'timeout': 10, u'look_for_keys': True, u'allow_agent': True})
    [81.2.255.191] Connected
    [pyinfra.api.state] Activating host: 81.2.255.191

--> Preparing operations...
    Loading: update_vps.py
    [pyinfra.api.operation] Adding operation, set([u'Apt/Update']), called @ update_vps.py:8, opLines=(0, 8), opHash=e7b37f7edef2ef843cea2aa2075da5d7f922fee3
    [81.2.255.191] Ready: update_vps.py

--> Proposed changes:
    Groups: kitakitsune_inventory / vps
    [81.2.255.191]   Operations: 1   Commands: 1   

--> Beginning operation run...
--> Starting operation: Apt/Update 
    [pyinfra.api.operations] Starting operation Apt/Update on 81.2.255.191
Traceback (most recent call last):
  File "src/gevent/greenlet.py", line 854, in gevent._gevent_cgreenlet.Greenlet.run
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/operations.py", line 129, in _run_server_op
    **executor_kwargs
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/host.py", line 126, in run_shell_command
    return self.executor.run_shell_command(self.state, self, *args, **kwargs)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 258, in run_shell_command
    put_file=put_file,
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/util.py", line 142, in get_sudo_password
    put_file(state, host, SUDO_ASKPASS_EXE, filename)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 427, in put_file
    _put_file(host, filename_or_io, remote_filename)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 381, in _put_file
    sftp.putfo(file_io, remote_location)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 714, in putfo
    with self.file(remotepath, "wb") as fr:
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 372, in open
    t, msg = self._request(CMD_OPEN, filename, imode, attrblock)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 813, in _request
    return self._read_response(num)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 865, in _read_response
    self._convert_status(msg)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 896, in _convert_status
    raise IOError(errno.EACCES, text)
IOError: [Errno 13] Permission denied
2020-07-26T18:31:40Z <Greenlet at 0x7fbd58e80160: _run_server_op(<pyinfra.api.state.State object at 0x7fbd58ec4a50>, 81.2.255.191, 'e7b37f7edef2ef843cea2aa2075da5d7f922fee3')> failed with IOError

--> An unexpected exception occurred:

  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 896, in _convert_status
    raise IOError(errno.EACCES, text)
IOError: [Errno 13] Permission denied
    [pyinfra_cli.exceptions]   File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra_cli/main.py", line 220, in main
    _main(*args, **kwargs)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra_cli/main.py", line 576, in _main
    run_ops(state, serial=serial, no_wait=no_wait)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/operations.py", line 351, in run_ops
    _run_single_op(state, op_hash)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/operations.py", line 317, in _run_single_op
    if not greenlet.get():
  File "src/gevent/greenlet.py", line 769, in gevent._gevent_cgreenlet.Greenlet.get
  File "src/gevent/greenlet.py", line 364, in gevent._gevent_cgreenlet.Greenlet._raise_exception
  File "src/gevent/greenlet.py", line 854, in gevent._gevent_cgreenlet.Greenlet.run
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/operations.py", line 129, in _run_server_op
    **executor_kwargs
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/host.py", line 126, in run_shell_command
    return self.executor.run_shell_command(self.state, self, *args, **kwargs)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 258, in run_shell_command
    put_file=put_file,
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/util.py", line 142, in get_sudo_password
    put_file(state, host, SUDO_ASKPASS_EXE, filename)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 427, in put_file
    _put_file(host, filename_or_io, remote_filename)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/pyinfra/api/connectors/ssh.py", line 381, in _put_file
    sftp.putfo(file_io, remote_location)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 714, in putfo
    with self.file(remotepath, "wb") as fr:
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 372, in open
    t, msg = self._request(CMD_OPEN, filename, imode, attrblock)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 813, in _request
    return self._read_response(num)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 865, in _read_response
    self._convert_status(msg)
  File "/home/bystrousak/.local/lib/python2.7/site-packages/paramiko/sftp_client.py", line 896, in _convert_status
    raise IOError(errno.EACCES, text)

    [pyinfra_cli.exceptions] IOError: [Errno 13] Permission denied


--> The full traceback has been written to pyinfra-debug.log
```

**Comment by Bystroushaak** [comment]
Exactly same operation work on older ubuntu, so maybe something changed on new ubuntu?

**Comment by Fizzadar** [comment]
@Bystroushaak does the SSH user have permission to write files to `/tmp`? I believe that's the issue here - the sudo password requires uploading a tiny script which then injects the password. An improvement here could be uploading to the users homedir, rather than assuming `/tmp` is write-able.

As an immediate/temporary fix you could either enable writing to `/tmp` or changing the `TEMP_DIR` config variable.

**Comment by Fizzadar** [comment]
I've changed the upload for the askpass script to the SSH users homedir, as this is a special case (https://github.com/Fizzadar/pyinfra/commit/1a90890c9beb0a539ee8376634e4f9bf8273ecfe). Should fix this issue, available in `1.1.dev0`!
